pkg/gadgets/traceloop/tracer/tracer.go

```diff
@@ -0,0 +1,436 @@
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	libseccomp "github.com/seccomp/libseccomp-golang"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target ${TARGET} -cc clang traceloop ./bpf/traceloop.bpf.c -- -I./bpf/ -I../../../${TARGET}
+
+var (
+	useNullByteLength        uint64 = uint64(C.USE_NULL_BYTE_LENGTH)
+	useRetAsParamLength      uint64 = uint64(C.USE_RET_AS_PARAM_LENGTH)
+	useArgIndexAsParamLength uint64 = uint64(C.USE_ARG_INDEX_AS_PARAM_LENGTH)
+	paramProbeAtExitMask     uint64 = uint64(C.PARAM_PROBE_AT_EXIT_MASK)
+)
+
+var (
+	syscallEventTypeEnter uint8 = uint8(C.SYSCALL_EVENT_TYPE_ENTER)
+	syscallEventTypeExit  uint8 = uint8(C.SYSCALL_EVENT_TYPE_EXIT)
+	syscallEventTypeCont  uint8 = uint8(C.SYSCALL_EVENT_TYPE_CONT)
+)
+
+// This should match traceloop.h define SYSCALL_ARGS.
+var syscallArgs uint8 = 6
+
+var syscalls map[string]syscallDeclaration
+
+type reader struct {
+	reader          *perf.Reader
+	previousHeadPos []uint64
+}
+
+type Tracer struct {
+	enricher gadgets.DataEnricher
+
+	spec *ebpf.CollectionSpec
+
+	// We cannot know when Attach()/Detach() will be called, so it is better to
+	// use lock to protect data structure which are not thought to be accessed
+	// in parallel.
+	objsLock  sync.Mutex
+	objs      traceloopObjects
+	enterLink link.Link
+	exitLink  link.Link
+
+	// Same comment than above, this map is designed to handled parallel access.
+	readers sync.Map
+}
+
+type syscallEvent struct {
+	timestamp uint64   `json:"timestamp,omitempty"`
+	typ       uint8    `json:"type,omitempty"`
+	contNr    uint8    `json:"contNr,omitempty"`
+	cpu       uint16   `json:"cpu,omitempty"`
+	id        uint16   `json:"id,omitempty"`
+	pid       uint64   `json:"pid,omitempty"`
+	comm      string   `json:"pcomm,omitempty"`
+	args      []uint64 `json:"args,omitempty"`
+	ret       uint64   `json:"ret,omitempty"`
+	mountNsID uint64   `json:"mountnsid,omitempty"`
+	retval    int      `json:"ret,omitempty"`
+}
+
+type syscallEventContinued struct {
+	timestamp uint64 `json:"timestamp,omitempty"`
+	index     uint8  `json:"index,omitempty"`
+	param     string `json:"params,omitempty"`
+}
+
+func NewTracer(enricher gadgets.DataEnricher) (*Tracer, error) {
+	t := &Tracer{
+		enricher: enricher,
+	}
+
+	spec, err := loadTraceloop()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	syscalls, err = gatherSyscalls()
+	if err != nil {
+		return nil, fmt.Errorf("error gathering syscall: %w", err)
+	}
+
+	// Fill the syscall map with specific syscall signatures.
+	syscallsMapSpec := spec.Maps["syscalls"]
+	for name, def := range syscallDefs {
+		nr, err := libseccomp.GetSyscallFromName(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get syscall number of %q: %w", name, err)
+		}
+
+		syscallsMapSpec.Contents = append(syscallsMapSpec.Contents, ebpf.MapKV{
+			Key:   uint64(nr),
+			Value: unsafe.Pointer(&def[0]),
```

Comment:
All values of the syscallsMapSpec.Contents variable are initialized to the same value because the pointer `unsafe.Pointer(&def[0])` has always the same value. When `spec.LoadAndAssign()` actually initializes the eBPF variable with `syscallsMapSpec.Contents`, it uses the same pointer for all.

Need to add:
```
		def := def
```

Checking with:
```
sudo bpftool map dump name syscalls
```